sdk/src/main/java/com/google/cloud/dataflow/sdk/testing/TestStream.java

```diff
@@ -0,0 +1,366 @@
+ * {@link Pipeline} after each method before it has completed and no more progress can be made by
+ * the {@link Pipeline}. A {@link PipelineRunner} must ensure that no more progress can be made in
+ * the {@link Pipeline} before advancing the state of the {@link TestStream}.
+ */
+public final class TestStream<T> extends PTransform<PBegin, PCollection<T>> {
+  private final List<Event<T>> events;
+  private final Coder<T> coder;
+
+  /**
+   * Create a new {@link TestStream.Builder} with no elements and watermark equal to {@link
+   * BoundedWindow#TIMESTAMP_MIN_VALUE}.
+   */
+  public static <T> Builder<T> create(Coder<T> coder) {
+    return new Builder<>(coder);
+  }
+
+  private TestStream(Coder<T> coder, List<Event<T>> events) {
+    this.coder = coder;
+    this.events = checkNotNull(events);
+  }
+
+  /**
+   * An incomplete {@link TestStream}. Elements added to this builder will be produced in sequence
+   * when the pipeline created by the {@link TestStream} is run.
+   */
+  public static class Builder<T> {
+    private final Coder<T> coder;
+    private final ImmutableList<Event<T>> events;
+    private final Instant currentWatermark;
+
+    private Builder(Coder<T> coder) {
+      this(coder, ImmutableList.<Event<T>>of(), BoundedWindow.TIMESTAMP_MIN_VALUE);
+    }
+
+    private Builder(Coder<T> coder, ImmutableList<Event<T>> events, Instant currentWatermark) {
+      this.coder = coder;
+      this.events = events;
+      this.currentWatermark = currentWatermark;
+    }
+
+    /**
+     * Adds the specified elements to the source with timestamp equal to the current watermark.
+     *
+     * @return A {@link TestStream.Builder} like this one that will add the provided elements
+     *         after all earlier events have completed.
+     */
+    @SafeVarargs
+    public final Builder<T> addElements(T element, T... elements) {
+      TimestampedValue<T> firstElement = TimestampedValue.of(element, currentWatermark);
+      @SuppressWarnings("unchecked")
+      TimestampedValue<T>[] remainingElements = new TimestampedValue[elements.length];
+      for (int i = 0; i < elements.length; i++) {
+        remainingElements[i] = TimestampedValue.of(elements[i], currentWatermark);
+      }
+      return addElements(firstElement, remainingElements);
+    }
+
+    /**
+     * Adds the specified elements to the source with the provided timestamps.
+     *
+     * @return A {@link TestStream.Builder} like this one that will add the provided elements
+     *         after all earlier events have completed.
+     */
+    @SafeVarargs
+    public final Builder<T> addElements(
+        TimestampedValue<T> element, TimestampedValue<T>... elements) {
+      checkArgument(
+          element.getTimestamp().isBefore(BoundedWindow.TIMESTAMP_MAX_VALUE),
+          "Elements must have timestamps before %s. Got: %s",
+          BoundedWindow.TIMESTAMP_MAX_VALUE,
+          element.getTimestamp());
+      for (TimestampedValue<T> multiElement : elements) {
```

Comment:
the name "multiElement" makes this confusing, because I thought looking in the code that it was something special, not just one of the elements). I'd suggest renamining the parameters to "firstElement" and "elements" then having the loop be "for element in elements"